New: no-const-reassign rule (fixes #2097)

jbrumwell opened the pull request:
The aim of this rule;

Variables that are declared as constants but are later reassigned, are no longer constant, and should be declared with let or var.

fixes: https://github.com/eslint/eslint/issues/2158

nzakas commented:
Thanks, please take a moment to read over our contribution guidelines: http://eslint.org/docs/developer-guide/contributing.html

Some of the things required:

1. One commit per pull request
2. Commit summary needs to be in the same format as the PR title you set

nzakas commented on conf/eslint.json:
```diff
@@ -136,7 +137,7 @@
         "new-cap": 2,
         "new-parens": 2,
-	"newline-after-var": 0,
+	    "newline-after-var": 0,
```
Misalignment?

nzakas commented on conf/eslint.json:
```diff
@@ -75,6 +75,7 @@
         "no-process-env": 0,
         "no-process-exit": 2,
         "no-proto": 2,
+        "no-reassign-constant": 0,
```
I think `no-const-reassign` would be better. Most rules with "assign" have that word last, and "const" makes it very clear we're dealing with `const` declarations.

nzakas commented on docs/rules/no-reassign-constant.md:
```diff
@@ -0,0 +1,46 @@
+# Disallow Reassignment of Constants (no-reassign-constant)
+
+Variables that are declared as constants but are later reassigned, are no longer constant, and should be declared with let or var.
```
This introduction needs to be longer, a few sentences at least. You should explain that reassigning to constants throws an error in ES6 in both strict and nonstrict mode and show an example.

nzakas commented on docs/rules/no-reassign-constant.md:
```diff
@@ -0,0 +1,46 @@
+
+## When Not to Use It
+
+If you don't want to be notified about variables that have been declared as constant, but have been reassigned you can turn this off.
```
Let's say if you aren't using ES6, then you don't need to use it.

nzakas commented on lib/rules/no-reassign-constant.js:
```diff
@@ -0,0 +1,141 @@
+/**
+ * @fileoverview Rule to flag reassignments of constants.
+ * @author Jason Brumwell
+ */
```
add copyright (see contribution guidelines)

nzakas commented on lib/rules/no-reassign-constant.js:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * Report a reference as invalid
+     *
+     * @param {String} name - the variables name
```
Lowercase - `string`

nzakas commented on lib/rules/no-reassign-constant.js:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * Check if a node is a unary delete expression
+     *
+     * @param {Node} node - The node to compare
```
`{ASTNode}`

nzakas commented on lib/rules/no-reassign-constant.js:
```diff
@@ -0,0 +1,141 @@
+            }
+
+            if (context.ecmaFeatures.blockBindings) {
+                checkScope(scope);
```
I'm not sure this is the best approach. Recursively searching through every scope is pretty inefficient. What you could do instead is listen for each `AssignmentExpression` and check to see if the lefthand side is a constant. That would dramatically cut down on the amount of work you're working doing for each file.

jbrumwell replied:
Thanks for the feedback, first time using eslint and the info you provided has helped a lot. As for the approach I originally was using ExpressionStatement to get UnaryExpressions and UpdateExpressions but I assumed the recursion would better performance. 

Here is the reasoning, please let me know if I'm mistaken, I assumed that at each level in the scope we are likely to have an assignment, update or update and possibly multiple. Since the expression doesn't expose the definition we need to take the variable name and loop through the variables in the scope. By looping through the defined variables instead of when they are changed, we decrease the number of times we have to loop through a scopes variables. Since we limit looping through references only to const, it should in theory work better?

nzakas replied:
Ah, I see. You may be right, I'll need to take a look at this again.

nzakas commented on tests/lib/rules/no-reassign-constant.js:
```diff
@@ -0,0 +1,34 @@
+var eslintTester = new ESLintTester(eslint);
+eslintTester.addRuleTest("lib/rules/no-reassign-constant", {
+    valid: [
+        {code: "const a = 'a';", ecmaFeatures: { blockBindings: true }},
```
You should add tests using `let` and `var` to ensure you're only flagging `const`

nzakas commented on lib/rules/no-reassign-constant.js:
```diff
@@ -0,0 +1,141 @@
+                };
+            }
+
+            if (context.ecmaFeatures.blockBindings) {
```
Seems like this should be the first line of the function, since there's no point in doing anything if the flag is false.

nzakas commented on lib/rules/no-reassign-constant.js:
```diff
@@ -0,0 +1,141 @@
+        "Program:exit": function() {
+            var scope = context.getScope();
+
+            if (scope.type === "global") {
```
What is this accomplishing?

jbrumwell replied:
For the global scope it does not seem to associate the references with the variables.

nzakas replied:
Of that's true, please file an issue with escope. I really don't want to be doing things like this.

jbrumwell commented:
Squashed commits and proper commit title used
Added tests for var, let and function scope
Adjusted program:exit handler to check if enabled first
Added additional information to the Rule documentation

Thanks for the feedback

nzakas commented on conf/eslint.json:
```diff
@@ -19,6 +19,7 @@
         "no-cond-assign": 2,
         "no-console": 2,
         "no-constant-condition": 2,
+        "no-constant-reassign": 0,
```
Still think this should be `no-const-reassign` as it relates only to `const` (and also, distinguishes it from `no-constant-condition`, which has nothing to do with `const`).

nzakas commented on docs/rules/README.md:
```diff
@@ -11,6 +11,7 @@ The following rules point out areas where you might have made mistakes.
 * [no-cond-assign](no-cond-assign.md) - disallow assignment in conditional expressions
 * [no-console](no-console.md) - disallow use of `console` (off by default in the node environment)
 * [no-constant-condition](no-constant-condition.md) - disallow use of constant expressions in conditions
+* [no-constant-reassign](no-constant-reassign.md) - disallow the reassignment of a variable declared as a constant (off by default)
```
Disallows reassignment of `const` declarations.

nzakas commented on docs/rules/no-constant-reassign.md:
```diff
@@ -0,0 +1,54 @@
+# Disallow Reassignment of Constants (no-constant-reassign)
+
+Variables that are declared as constants but are later reassigned, are no longer constant, and should be declared with let or var.
```
This section needs to show an example and mention `const`. The word "constant" is too ambiguous to stand alone.

nzakas commented on lib/rules/no-constant-reassign.js:
```diff
@@ -0,0 +1,142 @@
+        if (!scope.functionExpressionScope) {
+            for (var i = 0, l = variables.length; i < l; ++i) {
+                // skip function expression names
+                if (variables[i].eslintUsed) {
```
This skips every variable that was used? Why?

jbrumwell replied:
Thank you was a misunderstanding of the property meaning. Will push in a sec

jbrumwell commented:
changed name to no-const-reassign
altered documentation to avoid confusion over the term "constant"
added an additional example to the documentation 
removed bad code from the rule

NateEag commented:
Is this going to merge? The author closed it after improving it, which may have led to it falling through the cracks.

If this does merge, would it be worth adding an option to treat SNAKE_CAPPED vars as if they were declared const? That would make the rule useful for those of us who have to target browsers without `const`.

jbrumwell commented:
We couldn't devote any more time to it, but it is available as a plugin here https://github.com/colonyamerican/eslint-plugin-cah

clark800 commented:
+1 to adding this rule to eslint
